src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```diff
@@ -0,0 +1,373 @@
+        int extensionSize = record.getHeader().getWidth() -
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle);
+        sheetsSinkClient.extendDimension(spreadSheetId, spreadSheetName, sheetTitle, sheetId, extensionSize,
+          DimensionType.COLUMNS);
+        sheetsColumnCount.get(spreadSheetName).put(sheetTitle, record.getHeader().getWidth());
+      }
+
+      // 4. extend rows dimension if needed
+      if (sheetsContentShift.get(spreadSheetName).get(sheetTitle) >
+        sheetsRowCount.get(spreadSheetName).get(sheetTitle)) {
+        int minimumAddition = sheetsContentShift.get(spreadSheetName).get(sheetTitle) -
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle);
+        int extensionSize = Math.max(minimumAddition, googleSheetsSinkConfig.getMinPageExtensionSize());
+        sheetsSinkClient.extendDimension(spreadSheetId, spreadSheetName, sheetTitle, sheetId, extensionSize,
+          DimensionType.ROWS);
+        sheetsRowCount.get(spreadSheetName).put(sheetTitle,
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle) + extensionSize);
+      }
+
+      // 5. offer flatterned requests to queue
+      queueSemaphore.acquire();
+      try {
+        recordsQueue.offer(flatternedRowsRequest);
+      } finally {
+        queueSemaphore.release();
+      }
+
+      // 6. forcibly send requests to execution if the queue is exceeded
+      if (recordsQueue.size() >= googleSheetsSinkConfig.getRecordsQueueLength()) {
+        int counter = 0;
+        while (recordsQueue.size() >= googleSheetsSinkConfig.getRecordsQueueLength() && !stopSignal) {
+          if (counter > 0) {
+            Thread.sleep(1000);
+          }
+          formerScheduledExecutorService.submit(new TasksFormer(false)).get();
+          counter++;
+        }
+      }
+
+    } catch (ExecutionException | RetryException e) {
+      throw new InterruptedException(e.getMessage());
+    }
+  }
+
+  @Override
+  public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+    stopSignal = true;
+    // wait for scheduled task formers will be completed
+    formerScheduledExecutorService.shutdown();
+    formerScheduledExecutorService.awaitTermination(googleSheetsSinkConfig.getMaxFlushInterval() * 2,
+      TimeUnit.SECONDS);
+
+    // we should guarantee that at least one task former was called finally
+    try {
+      formerScheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+      formerScheduledExecutorService.submit(new TasksFormer(false)).get();
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    // wait for worker threads completion
+    writeService.shutdown();
+    writeService.awaitTermination(flushTimeout, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Task that forms batches of requests and schedules tasks for writing the batches.
+   */
+  public class TasksFormer implements Callable {
```

Comment:
create threading package and move the classes and other related logic (from methods possibly) there.

Reply:
added